core/src/main/java/io/questdb/cairo/wal/ApplyWal2TableJob.java

```diff
@@ -501,75 +510,76 @@ private void handleWalApplyFailure(TableToken tableToken, Throwable throwable, S
-                        final long rowsAdded = writer.commitWalTransaction(
-                                walPath,
-                                !dataInfo.isOutOfOrder(),
-                                dataInfo.getStartRowID(),
-                                dataInfo.getEndRowID(),
-                                dataInfo.getMinTimestamp(),
-                                dataInfo.getMaxTimestamp(),
-                                dataInfo,
-                                seqTxn,
-                                regulator
-                        );
-                        final long latency = microClock.getTicks() - start;
-                        long physicalRowCount = writer.getPhysicallyWrittenRowsSinceLastCommit();
-                        metrics.addApplyRowsWritten(rowCount, physicalRowCount, latency);
-                        walTelemetryFacade.store(WAL_TXN_DATA_APPLIED, writer.getTableToken(), walId, seqTxn, rowsAdded, physicalRowCount, latency);
-                        return rowCount;
-                    } else {
-                        // re-build wal transaction details
-                        return -2L;
+        WalTxnDetails txnDetails = writer.getWalTnxDetails();
+        final byte walTxnType = txnDetails.getWalTxnType(seqTxn);
+        switch (walTxnType) {
+            case DATA:
+                if (txnDetails.hasRecord(seqTxn)) {
+                    final long start = microClock.getTicks();
+                    walTelemetryFacade.store(WAL_TXN_APPLY_START, writer.getTableToken(), walId, seqTxn, -1L, -1L, start - commitTimestamp);
+                    writer.commitWalInsertTransactions(
+                            walPath,
+                            seqTxn,
+                            pressureControl
+                    );
+                    final long latency = microClock.getTicks() - start;
+                    long totalPhysicalRowCount = writer.getPhysicallyWrittenRowsSinceLastCommit();
+                    long lastCommittedSeqTxn = writer.getAppliedSeqTxn();
+                    lastCommittedRows = 0;
+                    for (long s = seqTxn; s <= lastCommittedSeqTxn; s++) {
+                        long walRowCount = txnDetails.getSegmentRowHi(s) - txnDetails.getSegmentRowLo(s);
+                        long commitPhRowCount = s == lastCommittedSeqTxn ? totalPhysicalRowCount : 0;
+                        metrics.addApplyRowsWritten(walRowCount, commitPhRowCount, latency);
+                        walTelemetryFacade.store(WAL_TXN_DATA_APPLIED, writer.getTableToken(), walId, s, walRowCount, commitPhRowCount, latency);
+                        lastCommittedRows += walRowCount;
                     }
+                    return (int) (writer.getAppliedSeqTxn() - seqTxn + 1);
+                } else {
+                    // re-build wal transaction details
+                    return RELOAD_TXN_DETAILS;
```

Comment:
untested from "cairo", if there a test somewhere else that excises this?

Reply:
It cannot happen anymore, an earlier call would fail with IndexOutOfRange, I removed the if check